lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -411,12 +411,18 @@ def is_satisfied_by(self, requirement, candidate):
         # NOTE: pre-releases is when there are several user requirements
         # NOTE: and one of them is a pre-release that also matches a
         # NOTE: transitive dependency of another requirement.
-        allow_pre_release = self._with_pre_releases or not (
-            requirement.ver == '*' or
-            requirement.ver.startswith('<') or
-            requirement.ver.startswith('>') or
-            requirement.ver.startswith('!=')
-        ) or self._is_user_requested(candidate)
+        # NOTE: Finally, if a concrete artifact is met, like a Git repository,
+        # NOTE: subdirs, a tarball URL, or a local dir or file etc,
+        # NOTE: pre-releases are allowed regardless of any other circumstances.
+        allow_pre_release = (
+            self._with_pre_releases or
+            candidate.is_concrete_artifact or not (
+                requirement.ver == '*' or
+                requirement.ver.startswith('<') or
+                requirement.ver.startswith('>') or
+                requirement.ver.startswith('!=')
+            ) or self._is_user_requested(candidate)
```

Comment:
I think we should clean up the (now) unused stuff that was added in https://github.com/ansible/ansible/pull/73416/files, and rename `_is_user_requested` method to something more accurate since it's now just handling a `type: galaxy` corner case.

Reply:
It's a small change, but I am less confident about backwards compatibility.

Reply:
@s-hertel looks like that entire commit can be reverted :thinking:

Reply:
Reverted. Marking this as resolved.

Reply:
Had to revert it back. It seems to still be necessary.

Reply:
Thinking more, we might need that `is_pinned` property after all. To facilitate removal of `_is_user_requested`.

Reply:
> @s-hertel looks like that entire commit can be reverted 🤔

UPD: So while working on this, just reverting the commit didn't work, which sent me off into the rabbit hole. I realized that we have a deeper problem with how we process pre-releases — if a requirement on any level of depth in the dependency tree has a pinned dependency that wan't taken into account. With the help of my pip maintainer  friend, we've confirmed that pip does not have the same problem — if a pinned dependency is a pre-release, it's allowed automatically in pip, without requiring a `--pre` which I regard as the correct behavior. With that, _I knew_ what bug we had but couldn't make up what's the difference with how pip does things. Luckily, Pradyun pointed out the pip's debugging reporter and I used it to figure out what the resolver does. He also made a comment along the lines of processing pre-releases in a different hook — `find_matches()`, not `is_satisfied_by()`! This all led took me to the right exploration path and I think I've finally figured it out: _we need to stop disregarding pinned versions from any requirements on the current round that are pinned!_

It looks like I nailed it, finally. This PR needs some love before getting out of Draft but it feels close to being ready.

@s-hertel feel free to run some testing of your own.

Reply:
This and more corner cases have been addressed. Marking this thread as resolved.